Internal Review: Sagemaker in airflow

yangaws opened the pull request:
Description:
1. Update aws_base_hook to support customized botocore configs
2. Update s3_hook with two more simple APIs, create_bucket and load_file_obj
3. Add sagemaker_hook with major sagemaker APIs including APIs to create and describe (training job, tuning job, transform job, model, endpoint-config, endpoint)
4. Add sagemaker operators including creating training job, tuning job, transform job, model, endpoint-config, endpoint. And a special one called deploy operator which combines model, endpoint-config and endpoint creation.
5. Add sagemaker sensors for sensing training job, tuning job, transform job, model, endpoint which might be time-consuming.
6. Add unit tests for all above. 

Tests:
1. Manually run all the unit tests
2. Run example DAGs with Pytorch (training and deployment), Tensorflow (training and transform), Mxnet (Tuning)
3. Since this branch is developed based on airflow 1.10 stable branch. I will run tests again before submitting final PR since that one would be based on airflow master.

yangaws committed:
Add training, tuning, transform, model, endpoint-config, endpoint operators and related sensors

yangaws committed:
Tail cloudwatch logs during training

yangaws committed:
Add evaluation for config

yangaws committed:
Add role expansion

ashb commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+        if '/' in role:
+            return role
+        else:
+            return self.iam_conn.get_role(RoleName=role)['Role']['Arn']
```
This should probably be in the AwsHook.

ashb commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+
+    def check_s3_url(self, s3url):
+        """
+        Check if a s3 url exists
```
There should be a blank line between the prose and the `:param` etc line. (I will only mention this once but it applies multiple times)

tomfaulhaber replied:
Edit: Check if an S3 URL exists

ashb commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+        :param log_group: The name of the log group.
+        :type log_group: str
+        :param streams: A list of the log stream names. The position of the stream in this list is
+        the stream number.
```
Continuation lines for params/types should be intended to render properly:

```

        :param streams: A list of the log stream names. The position of the stream in this list is
            the stream number. 
```

ashb commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+        positions = positions or {s: Position(timestamp=0, skip=0) for s in streams}
+        event_iters = [self.log_stream(log_group, s, positions[s].timestamp, positions[s].skip)
+                       for s in streams]
+        events = [next(s) if s else None for s in event_iters]
```
By making this a list we evaluate the entire `event_iters` at this point, making the yield in this function pointless.

I think both events and event_iters should be left as generator expressions:

And I think events might be https://docs.python.org/2.7/library/itertools.html#itertools.chain ?

ashb commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+        self.check_training_config(config)
+
+        response = self.conn.create_training_job(
+            **config)
```
Style nit: this should be on the previous line.

ashb commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+                else:
+                    positions[stream_names[idx]] = Position(timestamp=event['timestamp'], skip=1)
+        else:
+            sys.stdout.flush()
```
What's this for? I don't see us printing to stdout directly anywhere? Have I missed it?

tomfaulhaber replied:
@yangaws this is leftover from when we were printing '.' s on the output line while waiting for the first stream. You can take it out.

ashb commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+        """
+
+        return self.conn\
+            .describe_hyper_parameter_tuning_job(
```
Join with previous line please.

ashb commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+        running = True
+
+        while running:
+            time.sleep(check_interval)
```
@ashb Note to self: Come back to this once you've reviewed the rest. We might not want the loop in here, but to have it in a sensor instead.

ashb commented on airflow/contrib/operators/sagemaker_deploy_operator.py:
```diff
@@ -0,0 +1,164 @@
+       :param operation: Whether to create an endpoint or update an endpoint. Must be
+       one of 'create' and 'update'.
+       :type operation: str
+       **Example**:
```
Can you see how this renders in the docs:

`make -C docs html && open docs/_build/html/index.html`

(I guessed at the index filename from memory, so this might not be right.

ashb commented on airflow/contrib/operators/sagemaker_deploy_operator.py:
```diff
@@ -0,0 +1,164 @@
+               SageMakerDeployOperator(
+                   task_id='sagemaker_endpoint',
+                   config=request,
+                   region_name='us-west-2'
```
Please don't specify region_name by default - we pull that from the `extra` field of the connection ID by default.

ashb commented on airflow/contrib/operators/sagemaker_deploy_operator.py:
```diff
@@ -0,0 +1,164 @@
+            max_ingestion_time=self.max_ingestion_time
+        )
+        if not response['ResponseMetadata']['HTTPStatusCode'] \
+           == 200:
```
Join to previous line.

ashb commented on airflow/contrib/operators/sagemaker_model_operator.py:
```diff
@@ -0,0 +1,97 @@
+        self.config = config
+        self.region_name = region_name
+
+    def execute(self, context):
```
A lot of these operators feel similar. I'll need to take another look, but I feel that there might be a case for a BaseSageMakerOperator (or similar)

ashb reviewed (commented):
I've only given this a quick look over so far, and haven't read all the files yet.

tomfaulhaber commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+    def configure_s3_resources(self, config):
+        """
+        Evaluate the config of SageMaker operation and
+        execute related s3 operations
```
Edit: Extract the S3 operations from the configuration and execute them.

yangaws replied:
Updated

tomfaulhaber commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+
+    def check_training_config(self, training_config):
+        """
+        Check if a training config is valid
```
Edit: Check if a training configuration is valid

(generally we should use "configuration" in the prose parts of the doc strings  because it reads better.

yangaws replied:
Updated

tomfaulhaber commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+
+    def get_iam_conn(self, config=None):
+        """
+        Establish an AWS connection for retrieving iam roles during training
```
Edit: s/iam/IAM/

yangaws replied:
Updated

tomfaulhaber commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+        :return: A dict contains all the training job info
+        """
+
+        return self.conn\
```
Nit: the newline seems superfluous right here.

yangaws replied:
Updated

tomfaulhaber commented on airflow/contrib/hooks/sagemaker_hook.py:
```diff
@@ -0,0 +1,811 @@
+        return self.conn\
+            .describe_model(
+                ModelName=name)
+
```
Combine these onto one line and everywhere else we've got this weird multi-line construct, please.

tomfaulhaber commented on airflow/contrib/operators/sagemaker_deploy_operator.py:
```diff
@@ -0,0 +1,164 @@
+       Deploy a SageMaker model to an endpoint
+       This operator returns the ARN of the endpoint created in Amazon SageMaker
+       :param config:
+       The configuration necessary to create a model, an endpoint-config and
```
Where are the contents of this dictionary documented? The same for all of these Operators. We need complete docs, though I think it's ok to point to the AWS docs if the structures match.

tomfaulhaber commented on airflow/contrib/operators/sagemaker_deploy_operator.py:
```diff
@@ -0,0 +1,164 @@
+from airflow.exceptions import AirflowException
+
+
+class SageMakerDeployOperator(BaseOperator):
```
Why is this different than the `SageMakerCreateEndpointOperator`? I think that we should get rid of this and add optional model(s) and endpoint config to that one.

tomfaulhaber commented on airflow/contrib/operators/sagemaker_endpoint_config_operator.py:
```diff
@@ -0,0 +1,101 @@
+        self.region_name = region_name
+
+    def evaluate(self):
+        for variant in self.config['ProductionVariants']:
```
We should have commentary on why we're doing this everywhere that we do.

tomfaulhaber commented on airflow/contrib/operators/sagemaker_endpoint_config_operator.py:
```diff
@@ -0,0 +1,101 @@
+            % self.config['EndpointConfigName']
+        )
+        response = sagemaker.create_endpoint_config(self.config)
+        if not response['ResponseMetadata']['HTTPStatusCode'] \
```
```suggestion
        if response['ResponseMetadata']['HTTPStatusCode'] != 200:
```

yangaws replied:
Updated everywhere using this

tomfaulhaber commented on airflow/contrib/operators/sagemaker_model_operator.py:
```diff
@@ -0,0 +1,97 @@
+        )
+        response = sagemaker.create_model(self.config)
+        if not response['ResponseMetadata']['HTTPStatusCode'] \
+           == 200:
```
```
if response['ResponseMetadata']['HTTPStatusCode'] != 200:
```

If that feels too complicated to you, just factor out the status:
```
status = response['ResponseMetadata']['HTTPStatusCode']
if status != 200:
```

tomfaulhaber commented on airflow/contrib/operators/sagemaker_training_operator.py:
```diff
@@ -0,0 +1,143 @@
+            check_interval=self.check_interval,
+            max_ingestion_time=self.max_ingestion_time
+        )
+        if not response['ResponseMetadata']['HTTPStatusCode'] \
```
Again

tomfaulhaber commented on airflow/contrib/operators/sagemaker_tuning_operator.py:
```diff
@@ -0,0 +1,139 @@
+        self.max_ingestion_time = max_ingestion_time
+
+    def evaluate(self):
+        self.config['HyperParameterTuningJobConfig']['ResourceLimits']['MaxNumberOfTrainingJobs'] = \
```
Make sure that we can't have a null exception. 

It might be worth having a function `intify(dictionary, keys)` that you just call `intify(self.config, ['HyperParameterTuningJobConfig', 'ResourceLimits', 'MaxNumberOfTrainingJobs'])`

This would make these blocks a lot cleaner and let you handle null at each later. (I don't think that's a great function name, by the way.)

Is there anywhere that we might need to do intification (great word!) for all the elements of an array somewhere on the way down?

yangaws replied:
Add a base operator for all sagemaker operators.

Add a recursive function to parse integer fields in base operator.

tomfaulhaber commented on airflow/contrib/operators/sagemaker_tuning_operator.py:
```diff
@@ -0,0 +1,139 @@
+        self.check_interval = check_interval
+        self.max_ingestion_time = max_ingestion_time
+
+    def evaluate(self):
```
Should `evaluate` be implemented in a superclass and driven off a data table in conjunction with my comment below?

```python
self.integer_fields = [['HyperParameterTuningJobConfig', 'ResourceLimits', 'MaxNumberOfTrainingJobs'],
                       ['TrainingJobDefinition', 'ResourceConfig', 'InstanceCount'],
                       ['TrainingJobDefinition', 'ResourceConfig', 'VolumeSizeInGB'],
                       ['TrainingJobDefinition', 'StoppingCondition', 'MaxRuntimeInSeconds']]
```

If you wanted to be really clever, you could have a recursive structure:
```python
self.integer_fields = [['HyperParameterTuningJobConfig', 'ResourceLimits', 'MaxNumberOfTrainingJobs'],
                       ['TrainingJobDefinition', 
                           [['ResourceConfig', ['InstanceCount', 'VolumeSizeInGB']]],
                            ['StoppingCondition', 'MaxRuntimeInSeconds']]]]
```

But that's probably overkill :)

tomfaulhaber replied:
Also, `evaluate` isn't a very descriptive name of what this function does.

yangaws replied:
Now each sagemaker operator has its own integer_fields that can be evaluated.

tomfaulhaber reviewed (commented):
Generally looks good, but some things to clean up. A lot of the comments are doc and format nits, but there are a couple of more substantial things in here.

yangaws committed:
Address comments

yangaws committed:
Add base sagemaker operator

yangaws committed:
Fix unit tests

yangaws committed:
Address feedbacks from Ash in first PR

yangaws commented:
@tomfaulhaber 

Since the change is large, I've already sent first PR (about half of this) on 24th with addressing all the comments here. If you have further concerns, please continue reviewing this PR. I can update changes in later PRs.

THanks

yangaws committed:
More changes according to Ash

yangaws committed:
Update utc time conversion and unit tests

eslesar-aws commented on airflow/contrib/hooks/aws_hook.py:
```diff
@@ -150,15 +163,17 @@ def _get_credentials(self, region_name):
+                                config=config, verify=self.verify)
 
     def get_session(self, region_name=None):
         """Get the underlying boto3.session."""
```
"""Gets the underlying boto3 session."""

eslesar-aws commented on airflow/contrib/hooks/aws_hook.py:
```diff
@@ -171,7 +186,20 @@ def get_credentials(self, region_name=None):
         This contains the attributes: access_key, secret_key and token.
```
The session contains the following authentication attributes: ``acces_key``, ``secret_key``, and ``token``.

eslesar-aws commented on airflow/contrib/hooks/aws_hook.py:
```diff
@@ -171,7 +186,20 @@ def get_credentials(self, region_name=None):
+
+    def expand_role(self, role):
+        """
+        Expand an IAM role name to an IAM role ARN. If role is already an IAM ARN,
```
If the IAM role is a role name, get the Amazon Resource Name (ARN) for the role.

eslesar-aws commented on airflow/contrib/operators/sagemaker_endpoint_config_operator.py:
```diff
@@ -0,0 +1,71 @@
+    Create a SageMaker endpoint config
+    This operator returns The ARN of the endpoint config created in Amazon SageMaker
+
+    :param config: The configuration necessary to create an endpoint config
```
...to create an endpoint.

eslesar-aws replied:
Can we link to: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sagemaker.html#SageMaker.Client.create_endpoint_config

eslesar-aws commented on airflow/contrib/operators/sagemaker_endpoint_operator.py:
```diff
@@ -0,0 +1,143 @@
+    :type config: dict
+    :param aws_conn_id: The AWS connection ID to use.
+    :type aws_conn_id: str
+    :param wait_for_completion: if the operator should block
```
"Whether the operator should wait until the training job finishes.

eslesar-aws commented on airflow/contrib/operators/sagemaker_endpoint_operator.py:
```diff
@@ -0,0 +1,143 @@
+    :param wait_for_completion: if the operator should block
+        until training job finishes
+    :type wait_for_completion: bool
+    :param check_interval: if wait is set to be true, this is the time interval
```
If ``wait`` is set to ``True``, this is the time interval, in seconds, that this operation waits before polling the status of the training job.

eslesar-aws commented on airflow/contrib/operators/sagemaker_endpoint_operator.py:
```diff
@@ -0,0 +1,143 @@
+        in seconds which the operator will check the status of the training job
+    :type check_interval: int
+    :param max_ingestion_time: if wait is set to be true, the operator will fail
+        if the training job hasn't finish within the max_ingestion_time in seconds
```
If ``wait`` is set to ``True``, this operation fails if the training job doesn't finish within ``max_ingestion_time`` seconds.

eslesar-aws commented on airflow/contrib/operators/sagemaker_endpoint_operator.py:
```diff
@@ -0,0 +1,143 @@
+    :type check_interval: int
+    :param max_ingestion_time: if wait is set to be true, the operator will fail
+        if the training job hasn't finish within the max_ingestion_time in seconds
+        (Caution: be careful to set this parameters because training can take very long)
```
(Caution: make sure you set this parameter, because training can take a long time.) If you set this parameter to ``None`` it never times out.

eslesar-aws commented on airflow/contrib/operators/sagemaker_endpoint_operator.py:
```diff
@@ -0,0 +1,143 @@
+        Setting it to None implies no timeout.
+    :type max_ingestion_time: int
+    :param operation: Whether to create an endpoint or update an endpoint. Must be
+        one of 'create' and 'update'.
```
Must be either 'create or 'update'.

eslesar-aws commented on airflow/contrib/operators/sagemaker_training_operator.py:
```diff
@@ -0,0 +1,97 @@
+    :type config: dict
+    :param aws_conn_id: The AWS connection ID to use.
+    :type aws_conn_id: str
+    :param wait_for_completion: if the operator should block until training job finishes
```
Set to ``True`` to wait until the training job finishes.

eslesar-aws commented on airflow/contrib/operators/sagemaker_training_operator.py:
```diff
@@ -0,0 +1,97 @@
+    :type aws_conn_id: str
+    :param wait_for_completion: if the operator should block until training job finishes
+    :type wait_for_completion: bool
+    :param print_log: if the operator should print the cloudwatch log during training
```
Set to ``True`` to print the CloudWatch log during training.

eslesar-aws commented on airflow/contrib/operators/sagemaker_training_operator.py:
```diff
@@ -0,0 +1,97 @@
+    :type wait_for_completion: bool
+    :param print_log: if the operator should print the cloudwatch log during training
+    :type print_log: bool
+    :param check_interval: if wait is set to be true, this is the time interval
```
If ``wait`` is set to ``True``, the time interval, in seconds, that the operation waits to check the status of the training job.

eslesar-aws commented on airflow/contrib/operators/sagemaker_training_operator.py:
```diff
@@ -0,0 +1,97 @@
+    :param check_interval: if wait is set to be true, this is the time interval
+        in seconds which the operator will check the status of the training job
+    :type check_interval: int
+    :param max_ingestion_time: if wait is set to be true, the operator will fail
```
If ``wait`` is set to ``True``, the operation fails if the training job doesn't finish within ``max_ingestion_time`` seconds.
(Caution: make sure you set this parameter, because training can take a long time.) If you set this parameter to ``None``, the operation does not timeout.

eslesar-aws commented on airflow/contrib/operators/sagemaker_transform_operator.py:
```diff
@@ -0,0 +1,112 @@
+        The configuration necessary to create a SageMaker model, the default is none
+        which means the SageMaker model used for the SageMaker transform job already exists.
+        If given, it will be used to create a SageMaker model before creating
+        the SageMaker transform job
```
If a SageMaker model you want to use for the transform job does not already exist, provide a configuration to create a model. If the SageMaker model you want to use already exists, set this to ``None``.

eslesar-aws commented on airflow/contrib/operators/sagemaker_transform_operator.py:
```diff
@@ -0,0 +1,112 @@
+    :type model_config: dict
+    :param aws_conn_id: The AWS connection ID to use.
+    :type aws_conn_id: string
+    :param wait_for_completion: if the program should keep running until job finishes
```
Set to ``True`` to wait until the transform job finishes.

eslesar-aws commented on airflow/contrib/operators/sagemaker_transform_operator.py:
```diff
@@ -0,0 +1,112 @@
+    :type aws_conn_id: string
+    :param wait_for_completion: if the program should keep running until job finishes
+    :type wait_for_completion: bool
+    :param check_interval: if wait is set to be true, this is the time interval
```
If ``wait`` is set to ``True``, the time interval, in seconds, that this operation waits to check the status of the transform job.

eslesar-aws commented on airflow/contrib/operators/sagemaker_transform_operator.py:
```diff
@@ -0,0 +1,112 @@
+    :type check_interval: int
+    :param max_ingestion_time: if wait is set to be true, the operator will fail
+        if the transform job hasn't finish within the max_ingestion_time in seconds
+        (Caution: be careful to set this parameters because transform can take very long)
```
If ``wait`` is set to ``True``, the operation fails if the transform job doesn't finish within ``max_ingestion_time`` seconds.
(Caution: Make sure you set this parameter, because a transform job can take a long time.)

eslesar-aws commented on airflow/contrib/operators/sagemaker_tuning_operator.py:
```diff
@@ -0,0 +1,97 @@
+
+class SageMakerTuningOperator(SageMakerBaseOperator):
+    """
+    Initiate a SageMaker hyper-parameter tuning job.
```
...hyperparameter...

eslesar-aws commented on airflow/contrib/operators/sagemaker_tuning_operator.py:
```diff
@@ -0,0 +1,97 @@
+    :param max_ingestion_time: if wait is set to be true, the operator will fail
+        if the tuning job hasn't finish within the max_ingestion_time in seconds
+        (Caution: be careful to set this parameters because tuning can take very long)
+    :type max_ingestion_time: int
```
Same edits as previous sections

eslesar-aws commented on airflow/contrib/sensors/sagemaker_endpoint_sensor.py:
```diff
@@ -0,0 +1,61 @@
+class SageMakerEndpointSensor(SageMakerBaseSensor):
+    """
+    Asks for the state of the endpoint state until it reaches a terminal state.
+    If it fails the sensor errors, failing the task.
```
Polls the state of the endpoint until it reaches a terminal state.

eslesar-aws replied:
If it fails the sensor errors, the task fails.

comment: I don't really know what this means. What does it mean for an endpoint to reach a terminal state? Does that mean a termination of a response for one inference?

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -67,9 +69,36 @@ def get_bucket(self, bucket_name):
 
+    def create_bucket(self, bucket_name, region_name=None):
+        """
+        Creates a boto3.S3.Bucket object
```
Creates an Amazon S3 bucket.

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -67,9 +69,36 @@ def get_bucket(self, bucket_name):
+        """
+        Creates a boto3.S3.Bucket object
+
+        :param bucket_name: the name of the bucket
```
The

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -67,9 +69,36 @@ def get_bucket(self, bucket_name):
+
+        :param bucket_name: the name of the bucket
+        :type bucket_name: str
+        :param region__name: the name of the aws region
```
The name of the AWS region in which to create the bucket

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -67,9 +69,36 @@ def get_bucket(self, bucket_name):
+        :type bucket_name: str
+        :param prefix: a key prefix
+        :type prefix: str
+        :param delimiter: the delimiter marks key hierarchy.
```
The delimiter that marks key hierarchy. For example, for the prefix 'myTrainingJob/training/input/data', the delimiter is '/'.

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -227,6 +257,10 @@ def select_key(self, key, bucket_name=None,
             For more details about S3 Select parameters:
```
...parameters, see...

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -319,7 +362,7 @@ def load_string(self,
         This is provided as a convenience to drop a string in S3. It uses the
```
This is provided as a convenience to load a string into an S3 bucket.

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -377,3 +420,124 @@ def load_bytes(self,
+                      replace=False,
+                      encrypt=False):
+        """
+        Loads file object to S3
```
Loads a file object to S3.

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -377,3 +420,124 @@ def load_bytes(self,
+        """
+        Loads file object to S3
+
+        :param file_obj: file-like object to set as content for the key.
```
The file-like object to set as the content for the S3 key.

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -377,3 +420,124 @@ def load_bytes(self,
+        :type key: str
+        :param bucket_name: Name of the bucket in which to store the file
+        :type bucket_name: str
+        :param replace: A flag to decide whether or not to overwrite the key
```
A flag that indicates whether to overwrite the key if it already exists.

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -377,3 +420,124 @@ def load_bytes(self,
+        :param replace: A flag to decide whether or not to overwrite the key
+            if it already exists
+        :type replace: bool
+        :param encrypt: If True, the file will be encrypted on the server-side
```
If ``True``, S3 encrypts the file on the server, and the file is stored in encrypted form at rest in S3.

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -377,3 +420,124 @@ def load_bytes(self,
+
+            When it's specified as a full s3:// url, please omit source_bucket_name.
+        :type source_bucket_key: str
+        :param dest_bucket_key: The key of the object to copy to.
```
The key to copy the object to.

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -377,3 +420,124 @@ def load_bytes(self,
+            The convention to specify `dest_bucket_key` is the same
+            as `source_bucket_key`.
+        :type dest_bucket_key: str
+        :param source_bucket_name: Name of the S3 bucket where the source object is in.
```
The name of the S3 bucket that contains the source object.

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -377,3 +420,124 @@ def load_bytes(self,
+        """
+        :param bucket: Name of the bucket in which you are going to delete object(s)
+        :type bucket: str
+        :param keys: The key(s) to delete from S3 bucket.
```
...from the S3 bucket.

eslesar-aws commented on airflow/hooks/S3_hook.py:
```diff
@@ -377,3 +420,124 @@ def load_bytes(self,
+        :type bucket: str
+        :param keys: The key(s) to delete from S3 bucket.
+
+            When ``keys`` is a string, it's supposed to be the key name of
```
If ``keys`` is a string, it must be the key name of the single object to delete.

If ``keys`` is a list, it must be the list of the keys to delete.

yangaws committed:
Prepare second PR
